### PR TITLE
Provide GUID to zpool-reguid(8) with -g

### PR DESCRIPTION
**Pull request by 0mp**
### Motivation and Context

zpool-reguid can be used to change the GUID of a pool to a new, randomly generated one. Sometimes, the GUID needs to be set to a specific value, however. Currently, zpool-reguid does not offer a way to do that.

### Description

This commit extends the zpool-reguid(8) command with a -g flag, which allows the user to specify the GUID to set.

This change also adds some general tests for zpool-reguid(8).

Sponsored-by: Wasabi Technology, Inc.
Sponsored-by: Klara, Inc.

### How Has This Been Tested?

The patch adds appropriate tests to the test suite.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [x] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [x] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [x] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by 0mp** [comment]
I've pushed an update with the updated ABI files.

**Comment by clhedrick** [comment]
after a bit more thought: isn't the NFS fsid based on the guid of the file system? If so, I doubt this will help

File: cmd/zpool/zpool_main.c [hunk 1]
```diff
@@ -7595,19 +7595,32 @@ zpool_do_clear(int argc, char **argv)
...
 	zpool_handle_t *zhp;
 	int ret = 0;
 
 	/* check options */
-	while ((c = getopt(argc, argv, "")) != -1) {
+	while ((c = getopt(argc, argv, "g:")) != -1) {
 		switch (c) {
+		case 'g':
+			errno = 0;
+			guid = strtoull(optarg, &endptr, 10);
```
[the remark this answers is not in the document]
Comment (0mp) [reply cmd/zpool/zpool_main.c:7616]: Good idea! I've done that in the new revision.

File: man/man8/zpool-reguid.8 [hunk 2]
```diff
@@ -36,13 +37,23 @@
...
 .Ar pool
 .
 .Sh DESCRIPTION
 Generates a new unique identifier for the pool.
 You must ensure that all devices in this pool are online and healthy before
 performing this action.
 .
+.Bl -tag -width Ds
+.It Fl g Ar guid
+Set the GUID to the provided value.
```
[the remark this answers is not in the document]
Comment (0mp) [reply man/man8/zpool-reguid.8]: Done! Thanks!

**Review by robn — APPROVED** [review]

**Comment by tonyhutter** [comment]
I tested this out and was able to reguid to 0, which is reserved.  Fix:
```diff
diff --cc module/zfs/spa.c
index d50bc0bf9,50f5a0bf2..000000000
--- a/module/zfs/spa.c
+++ b/module/zfs/spa.c
@@@ -997,6 -997,6 +997,11 @@@ spa_change_guid(spa_t *spa, const uint6
  
        if (guidp != NULL) {
                guid = *guidp;
+               if (guid == 0) {
+                       error = SET_ERROR(EINVAL);
+                       goto out;
+               }
+
                if (spa_guid_exists(guid, 0)) {
                        error = SET_ERROR(EEXIST);
                        goto out;
```
Can you also rebase on master?

**Comment by 0mp** [comment]
> I tested this out and was able to reguid to 0, which is reserved. 

Good catch. Thank you for the patch. I've applied it, rebased the branch on top of recent master and pushed a new version.

**Review by behlendorf — APPROVED** [review]

**Review by tonyhutter — APPROVED** [review]

**Comment by 0mp** [comment]
Thanks!

**Pull request merged**